File: backend/src/Alife.Domain/Entities/EventPackage.cs
using Alife.Domain.Enums;

namespace Alife.Domain.Entities;

public sealed class EventPackageGovernancePolicyVersion
{
    public Guid Id { get; set; }
    public Guid? OrganisationId { get; set; }
    public string Version { get; set; } = string.Empty;
    public string SchemaVersion { get; set; } = "1";
    public string RulesJson { get; set; } = "{}";
    public EventPackageEnforcementMode EnforcementMode { get; set; } = EventPackageEnforcementMode.Off;
    public DateTime EffectiveFromUtc { get; set; }
    public DateTime? RetiredUtc { get; set; }
    public bool IsPublished { get; set; }
    public Guid PublishedByMemberId { get; set; }
    public DateTime PublishedUtc { get; set; }

    public Group? Organisation { get; set; }
    public Member PublishedByMember { get; set; } = null!;
    public ICollection<EventPackage> EventPackages { get; set; } = [];
}

public sealed class EventPackage
{
    public Guid Id { get; set; }
    public Guid EventId { get; set; }
    public EventPackageScopeType ScopeType { get; set; } = EventPackageScopeType.Event;
    public Guid? ScopeId { get; set; }
    public EventPackageCoverageMode CoverageMode { get; set; } = EventPackageCoverageMode.ExplicitOccurrences;
    public string CoveredOccurrenceIdsJson { get; set; } = "[]";
    public int Version { get; set; }
    public int EventPlanVersion { get; set; }
    public string PackageSchemaVersion { get; set; } = "1.0";
    public Guid GovernancePolicyVersionId { get; set; }
    public string GovernancePolicyVersion { get; set; } = string.Empty;
    public EventGovernanceTier GovernanceTier { get; set; }
    public EventPackageStatus Status { get; set; } = EventPackageStatus.Draft;
    public EventPackageApprovalValidity ApprovalValidityStatus { get; set; } = EventPackageApprovalValidity.NotDecided;
    public string ContentHash { get; set; } = string.Empty;
    public string SourceVectorHash { get; set; } = string.Empty;
    public string ManifestJson { get; set; } = "{}";
    public Guid? SupersedesPackageId { get; set; }
    public Guid GeneratedByMemberId { get; set; }
    public DateTime GeneratedUtc { get; set; }
    public Guid? SubmittedByMemberId { get; set; }
    public DateTime? SubmittedUtc { get; set; }
    public Guid ConcurrencyToken { get; set; } = Guid.NewGuid();

    public GroupEvent Event { get; set; } = null!;
    public EventOccurrence? ScopeOccurrence { get; set; }
    public EventPackageGovernancePolicyVersion GovernancePolicy { get; set; } = null!;
    public EventPackage? SupersedesPackage { get; set; }
    public Member GeneratedByMember { get; set; } = null!;
    public Member? SubmittedByMember { get; set; }
    public ICollection<EventPackageSourceReference> SourceReferences { get; set; } = [];
    public ICollection<EventPackageDecision> Decisions { get; set; } = [];
    public ICollection<EventPackageCondition> Conditions { get; set; } = [];
}

public sealed class EventPackageDecision
{
    public Guid Id { get; set; }
    public Guid EventPackageId { get; set; }
    public EventPackageDecisionType DecisionType { get; set; }
    public Guid ActorMemberId { get; set; }
    public string ReasonEn { get; set; } = string.Empty;
    public string ReasonZh { get; set; } = string.Empty;
    public DateTime DecidedUtc { get; set; }
    public string DecisionAuthoritySnapshotJson { get; set; } = "{}";
    public DateTime EffectiveUtc { get; set; }
    public DateTime? ExpiresUtc { get; set; }
    public Guid? RevokedByDecisionId { get; set; }
    public string? InvalidatedReasonCode { get; set; }
    public string RequestHash { get; set; } = string.Empty;

    public EventPackage EventPackage { get; set; } = null!;
    public Member ActorMember { get; set; } = null!;
    public EventPackageDecision? RevokedByDecision { get; set; }
    public ICollection<EventPackageCondition> WaivedConditions { get; set; } = [];
}

public sealed class EventPackageCondition
{
    public Guid Id { get; set; }
    public Guid EventPackageId { get; set; }
    public Guid? ReadinessTaskId { get; set; }
    public string TextEn { get; set; } = string.Empty;
    public string TextZh { get; set; } = string.Empty;
    public EventLifecycleGate AppliesToGate { get; set; }
    public string OwnerRoleRequirementKey { get; set; } = string.Empty;
    public DateTime DueUtc { get; set; }
    public EventPackageConditionStatus Status { get; set; } = EventPackageConditionStatus.Open;
    public DateTime? ExpiredUtc { get; set; }
    public Guid? WaivedByDecisionId { get; set; }
    public string? EvidenceReference { get; set; }
    public string? EvidenceReferenceHash { get; set; }
    public DateTime? EvidenceExpiresUtc { get; set; }
    public DateTime? EvidenceUnavailableUtc { get; set; }
    public Guid? SatisfiedByMemberId { get; set; }
    public DateTime? SatisfiedUtc { get; set; }
    public Guid? VerifiedByMemberId { get; set; }
    public DateTime? VerifiedUtc { get; set; }
    public Guid ConcurrencyToken { get; set; } = Guid.NewGuid();

    public EventPackage EventPackage { get; set; } = null!;
    public EventTask? ReadinessTask { get; set; }
    public EventPackageDecision? WaivedByDecision { get; set; }
    public Member? SatisfiedByMember { get; set; }
    public Member? VerifiedByMember { get; set; }
}

public sealed class EventPackageSourceReference
{
    public Guid Id { get; set; }
    public Guid EventPackageId { get; set; }
    public string ModuleCode { get; set; } = string.Empty;
    public string SubjectType { get; set; } = string.Empty;
    public Guid SubjectId { get; set; }
    public string SubjectVersion { get; set; } = string.Empty;
    public Guid? SourceDecisionId { get; set; }
    public DateTime? ValidUntilUtc { get; set; }
    public string DataClass { get; set; } = string.Empty;
    public bool RequiredForDecision { get; set; }
    public DateTime CapturedUtc { get; set; }

    public EventPackage EventPackage { get; set; } = null!;
}

public sealed class EventPackageApprovalDelegation
{
    public Guid Id { get; set; }
    public Guid OrganisationId { get; set; }
    public EventPackageDelegationScopeType ScopeType { get; set; }
    public Guid? ScopeId { get; set; }
    public string PermissionCode { get; set; } = "event.package.decide";
    public Guid DelegatedToMemberId { get; set; }
    public DateTime StartsUtc { get; set; }
    public DateTime ExpiresUtc { get; set; }
    public Guid GrantedByMemberId { get; set; }
    public DateTime GrantedUtc { get; set; }
    public Guid? RevokedByMemberId { get; set; }
    public DateTime? RevokedUtc { get; set; }
    public string? RevocationReasonEn { get; set; }
    public string? RevocationReasonZh { get; set; }
    public Guid ConcurrencyToken { get; set; } = Guid.NewGuid();

    public Group Organisation { get; set; } = null!;
    public Member DelegatedToMember { get; set; } = null!;
    public Member GrantedByMember { get; set; } = null!;
    public Member? RevokedByMember { get; set; }
}

File: cloudflare/alife-app/src/types/eventPackage.ts
import type { LocalizedText } from './eventComposition'

export type EventPackageScopeType = 'event' | 'occurrence'
export type EventPackageCoverageMode = 'explicitOccurrences' | 'planBoundSeriesWindow'
export type EventGovernanceTier = 'light' | 'standard' | 'enhanced'
export type EventPackageStatus = 'draft' | 'submitted' | 'returnedForAmendment' | 'rejected' | 'approvedWithConditions' | 'approved' | 'withdrawn' | 'superseded'
export type EventPackageApprovalValidity = 'notDecided' | 'active' | 'invalidated' | 'expired' | 'revoked'
export type LegacyEventPackageTransition = 'formalPackageRequired' | 'legacyReadOnlyPackage' | 'timeLimitedCompatibility' | 'safetyCriticalBlocked'
export type EventPackageDecisionType = 'approve' | 'approveWithConditions' | 'returnForAmendment' | 'reject' | 'revoke' | 'conditionWaiver'
export type EventPackageConditionStatus = 'open' | 'evidenceSubmitted' | 'verified' | 'rejected' | 'expired' | 'waived'
export type EventLifecycleGate = 'publish' | 'registration' | 'payment' | 'execute'

export type EventPackageModuleSummary = {
  moduleCode: string
  planStatus: string
  availability: 'available' | 'unavailable'
  sourceVersion: string
  blockers: LocalizedText[]
}

export type EventPackageManifest = {
  packageSchemaVersion: string
  eventId: string
  scopeType: EventPackageScopeType
  scopeId?: string | null
  coverageMode: EventPackageCoverageMode
  coveredOccurrenceIds: string[]
  eventPlanVersion: number
  governancePolicyVersion: string
  governanceTier: EventGovernanceTier
  legacyTransition: LegacyEventPackageTransition
  eventTitle: LocalizedText
  startUtc: string
  endUtc: string
  modules: EventPackageModuleSummary[]
  blockers: LocalizedText[]
  /** Optional while pre-1.0 stored manifests remain readable during rollout. */
  triggerReasons?: Array<{ code: string; message: LocalizedText }>
  requiredSpecialistDecisions?: string[]
  sections?: Array<{
    code: string
    title: LocalizedText
    status: 'ready' | 'attentionRequired' | 'notApplicable'
    items: LocalizedText[]
    moduleCodes: string[]
    blockers: LocalizedText[]
  }>
  warnings?: LocalizedText[]
}

export type EventPackage = {
  id: string
  eventId: string
  scopeType: EventPackageScopeType
  scopeId?: string | null
  coverageMode: EventPackageCoverageMode
  coveredOccurrenceIds: string[]
  version: number
  eventPlanVersion: number
  packageSchemaVersion: string
  governancePolicyVersion: string
  governanceTier: EventGovernanceTier
  status: EventPackageStatus
  approvalValidityStatus: EventPackageApprovalValidity
  contentHash: string
  sourceVectorHash: string
  manifest: EventPackageManifest
  sourceReferences: Array<{
    moduleCode: string
    subjectType: string
    subjectId: string
    subjectVersion: string
    sourceDecisionId?: string | null
    validUntilUtc?: string | null
    dataClass: string
    requiredForDecision: boolean
    capturedUtc: string
  }>
  decisions: Array<{
    id: string
    decisionType: EventPackageDecisionType
    actorMemberId: string
    reason: LocalizedText
    decidedUtc: string
    effectiveUtc: string
    expiresUtc?: string | null
    revokedByDecisionId?: string | null
    invalidatedReasonCode?: string | null
  }>
  conditions: Array<{
    id: string
    readinessTaskId?: string | null
    text: LocalizedText
    appliesToGate: EventLifecycleGate
    ownerRoleRequirementKey: string
    dueUtc: string
    status: EventPackageConditionStatus
    expiredUtc?: string | null
    evidenceReference?: string | null
    evidenceReferenceHash?: string | null
    evidenceExpiresUtc?: string | null
    evidenceUnavailableUtc?: string | null
    evidenceAvailable: boolean
    satisfiedByMemberId?: string | null
    satisfiedUtc?: string | null
    verifiedByMemberId?: string | null
    verifiedUtc?: string | null
    eTag: string
  }>
  supersedesPackageId?: string | null
  generatedByMemberId: string
  generatedUtc: string
  eTag: string
}

export type EventPackagePage = {
  items: EventPackage[]
  page: number
  pageSize: number
  totalCount: number
}

export type EventPackageDecisionRequest = {
  decisionType: Exclude<EventPackageDecisionType, 'revoke' | 'conditionWaiver'>
  reason: LocalizedText
  expiresUtc?: string | null
  conditions?: Array<{
    text: LocalizedText
    appliesToGate: EventLifecycleGate
    ownerRoleRequirementKey: string
    dueUtc: string
  }>
}

export type EventPublicationStatus = 'legacyImplicit' | 'draft' | 'published' | 'unpublished'
export type EventRegistrationStatus = 'legacyImplicit' | 'closed' | 'open'

export type EventLifecycle = {
  eventId: string
  publicationStatus: EventPublicationStatus
  publishedPackageId?: string | null
  publishedUtc?: string | null
  gateMode: 'off' | 'dryRun' | 'enforced'
  publishGateSatisfied: boolean
  reasonCodes: string[]
  eTag: string
  registrationStatus: EventRegistrationStatus
  registrationPackageId?: string | null
  registrationOpenedUtc?: string | null
  registrationGateMode: 'off' | 'dryRun' | 'enforced'
  registrationGateSatisfied: boolean
  registrationReasonCodes: string[]
  registrationETag: string
  executionStatus: 'notConfirmed' | 'confirmed' | 'invalidated'
  executionPackageId?: string | null
  executionConfirmedUtc?: string | null
  executionGateMode: 'off' | 'dryRun' | 'enforced'
  executionGateSatisfied: boolean
  executionReasonCodes: string[]
  executionETag: string
  paymentGateSatisfied: boolean
  paymentReasonCodes: string[]
  gates: Array<{
    gate: EventLifecycleGate
    enforcementMode: 'off' | 'dryRun' | 'enforced'
    scopeType: EventPackageScopeType
    scopeId?: string | null
    allowed: boolean
    requirementsSatisfied: boolean
    evaluatedUtc: string
    eventPlanVersion?: number | null
    eventPackageVersion?: number | null
    governancePolicyVersion?: string | null
    blockers: Array<{
      code: string
      message: LocalizedText
      responsibleRole: string
      nextAction: string
    }>
    warnings: Array<{
      code: string
      message: LocalizedText
      responsibleRole: string
      nextAction: string
    }>
  }>
}

export type EventPackageConditionResult = {
  condition: EventPackage['conditions'][number]
  lifecycle: EventLifecycle
}

export type EventPackageDiff = {
  fromPackageId: string
  fromVersion: number
  toPackageId: string
  toVersion: number
  hasMaterialChanges: boolean
  changes: Array<{
    field: string
    before?: string | null
    after?: string | null
    classification: 'cosmetic' | 'operational' | 'governanceCritical'
    affectedModuleCodes: string[]
  }>
}

export type EventPackageActorCapabilities = {
  eventId: string
  packageId: string
  canGenerate: boolean
  canSubmit: boolean
  canWithdraw: boolean
  canDecide: boolean
  canRevokeDecision: boolean
  canPublish: boolean
  canUnpublish: boolean
  canOpenRegistration: boolean
  canCloseRegistration: boolean
  canConfirmExecution: boolean
  canManageDelegations: boolean
  conditions: Array<{
    conditionId: string
    canSatisfy: boolean
    canVerify: boolean
    canWaive: boolean
  }>
}

export type EventPackageApprovalDelegation = {
  id: string
  organisationId: string
  scopeType: 'organisation' | 'event' | 'occurrence'
  scopeId?: string | null
  permissionCode: string
  delegatedToMemberId: string
  startsUtc: string
  expiresUtc: string
  grantedByMemberId: string
  grantedUtc: string
  revokedByMemberId?: string | null
  revokedUtc?: string | null
  revocationReason?: LocalizedText | null
  eTag: string
}
